content/en/blog/_posts/2022-03-31-service-ip-dynamic-and-static-allocation.md

```diff
@@ -0,0 +1,98 @@
+**Author:** Antonio Ojea
+
+
+Kubernetes Services are an abstract way to expose an application running on a set of Pods. Services
+has a `ClusterIP` that is virtual and allows to load-balance traffic across the different Pods. 
+
+## How Service ClusterIPs are allocated?
+
+A Service `ClusterIP` can be assigned:
+
+- dynamically, the cluster will automatically pick a free one within the configured Service IP
+  range.
+
+- statically, the user will set one IP within the configured Service IP range.
+
+Service `ClusterIP` are unique, hence, trying to create a Service with a `ClusterIP` that has already
+been allocated will return an error.
+
+## Why do I need to reserve Service Cluster IPs?
+
+Because you may want to have Services running in well-known IP addresses, so other components and
+users in the cluster can use them.
+
+The best example is the DNS Service for the cluster. Usually, people assign the 10th IP address in
+the Service IP range to the DNS service. Assuming our cluster has configured the Service IP range
+10.96.0.0/16 and we want our DNS Service IP to be 10.96.0.10, we'll have to create a Service like
+this:
+
+```yaml
+apiVersion: v1
+kind: Service
+metadata:
+  labels:
+    k8s-app: kube-dns
+    kubernetes.io/cluster-service: "true"
+    kubernetes.io/name: CoreDNS
+  name: kube-dns
+  namespace: kube-system
+spec:
+  clusterIP: 10.96.0.10
+  ports:
+  - name: dns
+    port: 53
+    protocol: UDP
+    targetPort: 53
+  - name: dns-tcp
+    port: 53
+    protocol: TCP
+    targetPort: 53
+  selector:
+    k8s-app: kube-dns
+  type: ClusterIP
+```
+
+but as we explained before, the 10.96.0.10 IP has not been reserved, if other Services are created
+before or in parallel with dynamic allocation, there is a chance they can allocate this IP, hence,
+we will not be able to create the DNS Service because it will fail with a conflict error.
+
+## How can I avoid Service ClusterIP conflicts?
+
+Since Kubernetes 1.24, a new Feature Gate `ServiceIPStaticSubrange` allows to use a different IP
+allocation strategy for Services, reducing the risk of collision.
+
+The `ClusterIP` range will be divided, based on the formula `min(max(16, cidrSize / 16), 256)`,
+described as _never less than 16 or more than 256 with a graduated step between them_.
+
+Dynamically IP assignment will use the upper band by default, once this has been exhausted it will
+use the lower range. This will allow users to use static allocations on the lower band with a low
+risk of collision.
+
+Example:
+
+- Service IP CIDR: 192.168.0.0/16
+- Range Size: 2^16 - 2 = 65534
+- Band Offset: min(max(16,65536/16),256) = min(4096,256) = 256
+- Static band start: 192.168.0.1
+- Static band ends: 192.168.1.0
+
+     ┌─────────────┬─────────────────────────────────────────────┐
+     │   static    │                    dynamic                  │
+     └─────────────┴─────────────────────────────────────────────┘
+
+     ◄────────────► ◄────────────────────────────────────────────►
```

Comment:
This should use a diagram in SVG format.

You can draw this yourself or use a tool to help. If you can, also use Kubernetes color, typefaces, etc.

See [Improve your documentation with mermaid.js diagrams](https://www.k8s.dev/blog/2021/12/01/improve-your-documentation-with-mermaid.js-diagrams/) for what I'd hope&mdash;as a co-author&mdash;is a thorough explanation of how to make that SVG starting from some Mermaid code.